plonky2/src/hash/field_merkle_tree.rs

```diff
@@ -0,0 +1,337 @@
+        let mut digests = Vec::with_capacity(num_digests);
+        let digests_buf = capacity_up_to_mut(&mut digests, num_digests);
+        let mut digests_buf_pos = 0;
+
+        let mut cap = vec![];
+        let dummy_leaves = vec![vec![F::ZERO]; 1 << cap_height];
+        leaves.push(dummy_leaves);
+        for window in leaves.windows(2) {
+            let cur = &window[0];
+            let next = &window[1];
+
+            let cur_leaf_len = cur.len();
+            let next_cap_len = next.len();
+            let next_cap_height = log2_strict(next_cap_len);
+
+            leaf_heights.push(log2_strict(cur_leaf_len));
+
+            let num_tmp_digests = 2 * (cur_leaf_len - next_cap_len);
+
+            if cur_leaf_len == leaves_len {
+                // The bottom leaf layer
+                cap = Vec::with_capacity(next_cap_len);
+                let tmp_cap_buf = capacity_up_to_mut(&mut cap, next_cap_len);
+                fill_digests_buf::<F, H>(
+                    &mut digests_buf[digests_buf_pos..(digests_buf_pos + num_tmp_digests)],
+                    tmp_cap_buf,
+                    &cur[..],
+                    next_cap_height,
+                );
+            } else {
+                // The rest leaf layers
+                let new_leaves: Vec<Vec<F>> = cap
+                    .iter()
+                    .enumerate()
+                    .map(|(i, cap_hash)| {
+                        let mut new_hash = Vec::with_capacity(NUM_HASH_OUT_ELTS + cur[i].len());
+                        new_hash.extend(&cap_hash.to_vec());
+                        new_hash.extend(&cur[i]);
+                        new_hash
+                    })
+                    .collect();
+                cap.clear();
+                cap.reserve_exact(next_cap_len);
+                let tmp_cap_buf = capacity_up_to_mut(&mut cap, next_cap_len);
+                fill_digests_buf::<F, H>(
+                    &mut digests_buf[digests_buf_pos..(digests_buf_pos + num_tmp_digests)],
+                    tmp_cap_buf,
+                    &new_leaves[..],
+                    next_cap_height,
+                );
+            }
+
+            unsafe {
+                cap.set_len(next_cap_len);
+            }
+
+            digests_buf_pos += num_tmp_digests;
+        }
+
+        unsafe {
+            // SAFETY: `fill_digests_buf` and `cap` initialized the spare capacity up to
+            // `num_digests` and `len_cap`, resp.
+            digests.set_len(num_digests);
+        }
+
+        // remove dummy leaves
+        leaves.pop();
+
+        Self {
+            leaves,
+            digests,
+            cap: MerkleCap(cap),
+            leaf_heights,
+        }
+    }
+
+    /// Create a Merkle proof from a leaf index.
+    pub fn open_batch(&self, leaf_index: usize) -> MerkleProof<F, H> {
+        let mut digests_buf_pos = 0;
+        let leaves_cap_height = log2_strict(self.leaves[0].len());
```

Comment:
Extremely nit, but it confused me for a bit since the base layer isn't a cap, unlike all other layers.

Reply:
Done.